overlord/hookstate/processrunner.go

```diff
@@ -0,0 +1,69 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package hookstate
+
+import (
+	"bytes"
+	"os/exec"
+)
+
+type ProcessRunner interface {
+	// Kill kills the running process (if any)
+	Kill() error
+
+	// Wait waits for the running process to complete, and returns combined
+	// stdout/stderr.
+	Wait() ([]byte, error)
+}
+
+type execProcessRunner struct {
+	command *exec.Cmd
+	buffer  *bytes.Buffer
+}
+
+// newProcessRunner starts a new process with the requested parameters and
+// returns its ProcessRunner.
+func newProcessRunner(name string, args ...string) (ProcessRunner, error) {
```

Comment:
It feels like there's way too much indirection to do something quite straightforward. We have the exactly the same functionality available in exec.Cmd itself.

I bet you can comfortably kill all of that, including even doStartCommand (which is unused!), ProcessRunner, newProcessRunner, and all its friends, and replace all of it by simply:
var buf bytes.Buffer
cmd := exec.Command("snap", "run", ...)
cmd.Stdin = &buf
cmd.Stdout = &buf
cmd.Start()
...

Inside `snapRun` itself (old `doRunHook`).

For testing, we have a good mechanism for mocking actual commands (executables).